### PR TITLE
[c-ares] Fix cross building

### PR DESCRIPTION
**Pull request by anton-danielsson**
Add option to build tools or not. Defaults to false when cross building and true otherwise.
Add patch to that adds CARES_BUILD_TOOLS on older versions (1.14.0)
    
This enables the package to be used in iOS and Android builds.

Specify library name and version:  **c-ares/all**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
Usually we add an option (others cross compilation can produce these executables).
This option could be `tools`.
Then you could replace your modification by `self._cmake.definitions["CARES_BUILD_TOOLS"] = self.options.tools`
And finally you should modify `package_info()`:

```python
        if self.options.tools:
            bin_path = os.path.join(self.package_folder, "bin")
            self.output.info("Appending PATH environment variable: {}".format(bin_path))
            self.env_info.PATH.append(bin_path)
```

**Comment by SpaceIm** [comment]
Please consider https://github.com/DiracResearch/conan-center-index/pull/1

**Comment by anton-dirac** [comment]
Good points @SpaceIm. I have applied the changes you suggested.
I made one slight adjustment to default the `tools` option to `False` if cross building, otherwise `True`.

FYI: I have applied for the EAP thing but I'm not in yet.
As soon as the CI system builds the PR I will make sure to check for any failures :)

File: recipes/c-ares/all/conanfile.py [hunk 1]
```diff
@@ -30,6 +37,9 @@ def _build_subfolder(self):
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        # Default build tools to false if cross building, otherwise to true.
+        if self.options.tools == None:
+            self.options.tools = not tools.cross_building(self.settings)
```
Comment (SpaceIm) [inline recipes/c-ares/all/conanfile.py:42]: Not sure if we should not delegate this responsability to consumers (coherent with others recipes).

Comment (anton-dirac) [reply recipes/c-ares/all/conanfile.py:42]: I just think that options should have good defaults and in almost all situations I think you are not interested to have these tools for the host environment when cross building. The recipe still always you to set the option if the default is not what you want.

I'm planing to contribute similar things for other recipes (like protobuf with it's protoc tool). So I guess it good to take make a decision here and go with it :)

Comment (SpaceIm) [reply recipes/c-ares/all/conanfile.py:42]: Why not. Let's wait opinion of others contributors.

Anyway, you don't need `if self.options.tools == None:`

Comment (anton-dirac) [reply recipes/c-ares/all/conanfile.py:42]: I think the `== None` is needed. See https://github.com/conan-io/conan/issues/3519.
The "pattern" is taken from the docs: https://docs.conan.io/en/latest/reference/conanfile/attributes.html#default-options
```py
    def config_options(self):
        if self.options.some_option == None:
            if self.settings.os == 'Android':
                self.options.some_option = True
            else:
                self.options.some_option = False
```
I noted that `conan create` and `conan install` behaves differently.
With `conan install` the None check is needed but not with `conan create`. Weird..

**Comment by anton-dirac** [comment]
Sorry, I don't understand what the errors are.
Is it that `create_stderr.txt` is not empty?
Like:
```
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY


CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
```
I think that's unrelated the this PR and is more likely caused by different CMake versions on the build agents.
(Warning cased by cmake policy stuff)
It can be fixed though, I just need to understand if that's the problem or not.
Could be that I'm completely blind and can't see the real problem...

**Comment by dmn-star** [comment]
This has always been my guess that the (wrong) stdout/stderr redirections/interceptions cause CCI to mark the successful build as failed. @jgsogo 

@anton-dirac Try to close the PR and open it again after 15 seconds, maybe the "errors" are gone.

**Comment by anton-dirac** [comment]
@dmn-star I have no idea why the build fails...
I have read the output logs over and over again and they look fine...

**Comment by dmn-star** [comment]
> @dmn-star I have no idea why the build fails...
> I have read the output logs over and over again and they look fine...

The CCI black box testing is always so time consuming. :-)  I'll try step by step. #4553

**Comment by dmn-star** [comment]
All of this is really interesting. 🤔

**Review by prince-chrismc — DISMISSED** [review]

**Review by jgsogo — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Comment by anton-danielsson** [comment]
Reviewers: 
I'm actually not 100% happy with this PR yet.
If option tools is enabled the iOS build will still fail with:
```
CMake Error at source_subfolder/src/tools/CMakeLists.txt:19 (INSTALL):
  INSTALL TARGETS given no BUNDLE DESTINATION for MACOSX_BUNDLE executable
  target "ahost".
CMake Error at source_subfolder/src/tools/CMakeLists.txt:36 (INSTALL):
  INSTALL TARGETS given no BUNDLE DESTINATION for MACOSX_BUNDLE executable
  target "adig".
CMake Error at source_subfolder/src/tools/CMakeLists.txt:53 (INSTALL):
  INSTALL TARGETS given no BUNDLE DESTINATION for MACOSX_BUNDLE executable
  target "acountry".
```
I think the only way to fix this is to patch the CMakeLists file in c-ares.
However, I see no reason to build these executables for iOS (or Android).
Would it be OK to just skip tools if iOS or Android?
If not I think I want to add a patch for c-ares.

**Comment by jgsogo** [comment]
It is totally fine. Someone can take care of them in the future if those tools are needed in a ios/android scenario. We cannot fix everything in each PRs, let's go step by step!

Every contribution helps to improve C++ ecosystem.

Thanks a lot!

**Comment by SpaceIm** [comment]
No because it will require to patch c-ares forever. I prefer to wait that conan supports setting default option value with `tools.cross_building(self.settings)`.

**Comment by anton-danielsson** [comment]
Ok, makes sense. So merge this and I might create a new PR in the future depending on how other things pan out?
I might also make a PR to upstream c-ares to fix the install problem. I agree that maintaining patches is a nightmare..

**Comment by anton-danielsson** [comment]
FYI: It was a one-liner fix in c-ares https://github.com/c-ares/c-ares/pull/395
With some luck it will be included in the next release :)

**Comment by SpaceIm** [comment]
Nice, it looks good to me to add a patch in all current versions of c-ares in CCI.

**Comment by anton-danielsson** [comment]
@SpaceIm Do you mean like: https://github.com/conan-io/conan-center-index/pull/4492/commits/65313e9f7c108fd834602f17d0688a72f2aaf257?
BTW, is there a magic trick to avoid having to create different patches for different versions?

**Comment by SpaceIm** [comment]
> 
> 
> @SpaceIm Do you mean like: [65313e9](https://github.com/conan-io/conan-center-index/commit/65313e9f7c108fd834602f17d0688a72f2aaf257)?
> BTW, is there a magic trick to avoid having to create different patches for different versions?

Yes and it depends.
If we must live with a patch forever, `tools.replace_in_file()` is your friend, otherwise patch files are preferred. Here you can't have the same patch for all versions since lines numbers are slightly different, and `tools.replace_in_file()` is not a good solution since it will be fixed upstream.

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by dmn-star — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**